### ISSUE TITLE
Preview 7: Flickery page transitions

### ISSUE THREAD

**Issue by mikegoatly** [AvaloniaUI/Avalonia#11167]
**Describe the bug**
Since upgrading from v11 preview 6 to preview 7 I've noticed that page transitions are very flickery:

https://user-images.githubusercontent.com/4577868/234988210-0d85f87d-a32a-49f9-be26-1d95a60245fd.mp4

The video is from a quick repro I put together based on the existing routing tutorial: https://github.com/mikegoatly/AvaloniaExperiments

I've tried both `CrossFade` and `PageSlide` transitions and the flickering is apparent in both.

**To Reproduce**

Clone the sample from https://github.com/mikegoatly/AvaloniaExperiments, or create a new project that uses a `RoutedViewHost`:

``` axaml
<rxui:RoutedViewHost Grid.Row="0" Router="{Binding Router}">
	<rxui:RoutedViewHost.DefaultContent>
		<TextBlock Text="Default content"
						 FontSize="40"
                         HorizontalAlignment="Center"
                         VerticalAlignment="Center" />
	</rxui:RoutedViewHost.DefaultContent>
	<rxui:RoutedViewHost.ViewLocator>
		<!-- See AppViewLocator.cs section below -->
		<app:AppViewLocator />
	</rxui:RoutedViewHost.ViewLocator>
</rxui:RoutedViewHost>
```

Switch between views and the animations flicker.

**Expected behavior**
The page transitions should be smooth.

**Desktop (please complete the following information):**

- OS: Windows, also tried in Linux with DRM
- Version 11.0.0-preview7

**Comment by mikegoatly** [comment]
This is what it looks like if I downgrade to preview 6:

https://user-images.githubusercontent.com/4577868/234991002-3e578cbd-991e-448d-a05c-d24b4cc4e451.mp4

**Comment by grokys** [comment]
Seems to have been caused by 5f52122bba4487ba948b070342979fda0a2d1f2f (#10792)

**Comment by grokys** [comment]
In addition to this problem, by slowing down the animation I can see that it's not actually cross-fading but fading one control out and the other in:

https://user-images.githubusercontent.com/1775141/235181398-21a1abb0-f338-4727-8ff1-9db3b7f1a74f.mp4

**Comment by maxkatz6** [comment]
> In addition to this problem, by slowing down the animation I can see that it's not actually cross-fading but fading one control out and the other in:

It's "expected", as our transitions work in an order "from" to "to". As well as we can't have two contents at the same time in the content presenter.  See https://github.com/AvaloniaUI/Avalonia/blob/f456ea21c558c1dfa82bcaf6c2a282b320188f4b/src/Avalonia.Controls/TransitioningContentControl.cs#L98-L109

**Comment by kekekeks** [comment]
>we can't have two contents at the same time in the content presenter
Well, it means that `TransitioningContentControl` needs to be changed to have two ContentPresenter instances.

**Comment by llfab** [comment]
> > we can't have two contents at the same time in the content presenter
> 
> Well, it means that `TransitioningContentControl` needs to be changed to have two ContentPresenter instances.

If someone really want the TCC to be able to do both (control1 out, then control2 in AND control1 out while control2 in). Not sure if that is a priority. Everything both the control and the anims (like CrosFade) had been able so far is to do one after another. I would consider fixing the issues a priority and extending its capabilities  to have a lower prio (allthough yes, fading one control into another is indeed a often required use case).

**Comment by grokys** [comment]
I agree that fixing TCC to have two ContentPresenters is less important than fixing the flickering. Fixing the flickering is next on my list of issues to fix, so I hope to fix it next week.

**Comment by llfab** [comment]
> I've got a fix for most of the flickering on the `fixes/11167-pagetransition-flicker` branch, but I need to wait for @kekekeks to open a PR for some incoming changes before I finish it off and tidy it up.

Thanks for that timely fix. I will try asap...

**Comment by llfab** [comment]
There will not be a NuGet for ```fixes/11167-pagetransition-flicker``` right. Only once this is a PR, correct?

**Comment by grokys** [comment]
Actually, I've got a fix that I can open a PR for now. Turns out that if we fix `TransitioningContentControl` to have two content presenters and transition between them correctly, then the problems go away.

**Comment by llfab** [comment]
> Actually, I've got a fix that I can open a PR for now. Turns out that if we fix `TransitioningContentControl` to have two content presenters and transition between them correctly, then the problems go away.

For the old TCC I implemented my own CrossFade to support transitioning to / from null content. As you said, originally to "from" was faded out first and when done the "to" was faded in. I cloned CrossFade and made it check if the to/from was null and kind of skipped the to/from animation. I am using the TCC often for fading in dialog layers which makes to/from = null a standard case. From what you describe, to/from = null should now work out of the box, right?
